src/BlockTypes/AddToCartForm.php

```diff
@@ -58,10 +50,460 @@ protected function render( $attributes, $content, $block ) {
 			esc_attr( $classes_and_styles['classes'] ),
 			esc_attr( $classname ),
 			esc_attr( $classes_and_styles['styles'] ),
-			$product
+			$add_to_cart_form
 		);
 	}
 
+	/**
+	 * Return the Add to Cart Form for a given product.
+	 *
+	 * @param \WC_Product $product Product object.
+	 *
+	 * @return string
+	 */
+	protected function add_to_cart_form( $product ) {
+		$product_type = $product->get_type();
+
+		switch ( $product_type ) {
+			case 'simple':
+				$add_to_cart_form = $this->add_simple_product_to_cart( $product );
+				break;
+			case 'variable':
+				$add_to_cart_form = $this->add_variable_product_to_cart( $product );
+				break;
+			case 'grouped':
+				$add_to_cart_form = $this->add_grouped_product_to_cart( $product );
+				break;
+			case 'external':
+				$add_to_cart_form = $this->add_external_product_to_cart( $product );
+				break;
+			default:
+				$add_to_cart_form = '';
+				break;
+		}
+
+		ob_start();
```

Comment:
> Is it necessary to buffer here as well, since this is essentially calling those other methods that already contain buffers?

Yes. Buffering was used here in this particular case to account for the last `do_action( 'woocommerce_after_add_to_cart_form' );`: it wouldn't be possible to trigger it if the `$add_to_cart_form` was immediately returned instead.

According to [PHP's documentation](https://www.php.net/ob_start), output buffers are stackable; If multiple output callback functions are active, the output is filtered sequentially through each in nesting order. IMO, we can safely stick with this approach.